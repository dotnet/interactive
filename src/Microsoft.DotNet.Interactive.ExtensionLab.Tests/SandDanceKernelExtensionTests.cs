// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;

using Assent;

using FluentAssertions;

using Microsoft.DotNet.Interactive.Formatting;
using Microsoft.DotNet.Interactive.Formatting.TabularData;
using Xunit;

namespace Microsoft.DotNet.Interactive.ExtensionLab.Tests
{
    public class SandDanceKernelExtensionTests : IDisposable
    {
        private readonly Configuration _configuration;

        public SandDanceKernelExtensionTests()
        {
            _configuration = new Configuration()
                .SetInteractive(Debugger.IsAttached)
                .UsingExtension("json");
        }

        [Fact]
        public async Task it_configures_preferred_mimeTypes()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();

            await kernelExtension.OnLoadAsync(kernel);

            var mimetypes = Formatter.GetPreferredMimeTypesFor(typeof(SandDanceDataExplorer)).Distinct();

            mimetypes
                .Should().BeEquivalentTo(HtmlFormatter.MimeType, TabularDataResourceFormatter.MimeType);
        }
        
        [Fact]
        public async Task it_registers_html_formatter()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();

            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };

            var formatted = data.ExploreWithSandDance().ToDisplayString(HtmlFormatter.MimeType);

            formatted.Should().Contain("configureRequireFromExtension('SandDance','1.0.0')(['SandDance/sanddanceapi'], (sandDance) => {");
        }

        [Fact]
        public async Task it_registers_tabularResource_formatter()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();

            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };


            var formattedValue = data.ExploreWithSandDance().ToDisplayString(TabularDataResourceFormatter.MimeType);
            formattedValue.Should().Contain("\"profile\": \"tabular-data-resource\"");
        }

        [Fact]
        public async Task it_is_formatted_as_multiple_mimeTypes()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();

            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };


            var formattedValues = FormattedValue.FromObject(data.ExploreWithSandDance());
            formattedValues.Select(fv => fv.MimeType)
                .Should()
                .BeEquivalentTo(HtmlFormatter.MimeType, TabularDataResourceFormatter.MimeType);
        }

        [Fact]
        public async Task widget_code_generation_is_not_broken()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new NteractKernelExtension();

            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };


            var html = data.ExploreWithSandDance().ToDisplayString(HtmlFormatter.MimeType);

            this.Assent(html.FixedGuid());
        }

        [Fact]
        public async Task it_can_load_script_from_the_extension()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();

            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };


            var formatted = data.ExploreWithSandDance().ToDisplayString(HtmlFormatter.MimeType);

            formatted.Should().Contain("configureRequireFromExtension");
        }

        [Fact]
        public async Task it_checks_and_load_require()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();
            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };

            var explorer = data.ExploreWithSandDance();
            explorer.LibraryUri = new Uri("https://a.cdn.url/script.js");
            var formatted = explorer.ToDisplayString(HtmlFormatter.MimeType);

            formatted.Should()
                .Contain("if ((typeof(require) !==  typeof(Function)) || (typeof(require.config) !== typeof(Function)))")
                .And
                .Contain("require_script.onload = function()");
        }

        [Fact]
        public async Task it_can_loads_script_from_uri()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();
            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };

            var explorer = data.ExploreWithSandDance();
            explorer.LibraryUri = new Uri("https://a.cdn.url/script.js");
            var formatted = explorer.ToDisplayString(HtmlFormatter.MimeType);

            formatted.Should().Contain("require.config(");
        }

        [Fact]
        public async Task it_can_loads_script_from_uri_and_specify_context()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();
            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };

            var explorer = data.ExploreWithSandDance();
            explorer.LibraryUri = new Uri("https://a.cdn.url/script.js");
            explorer.LibraryVersion = "2.2.2";
            var formatted = explorer.ToDisplayString(HtmlFormatter.MimeType);

            formatted.Should().Contain("'context': '2.2.2'");
        }

        [Fact]
        public async Task uri_is_quoted()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();
            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };

            var explorer = data.ExploreWithSandDance();
            explorer.LibraryUri = new Uri("https://a.cdn.url/script.js");
            var formatted = explorer.ToDisplayString(HtmlFormatter.MimeType);

            formatted.Should().Contain("'https://a.cdn.url/script'");
        }

        [Fact]
        public async Task uri_extension_is_removed()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();
            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };

            var explorer = data.ExploreWithSandDance();
            explorer.LibraryUri = new Uri("https://a.cdn.url/script.js");
            var formatted = explorer.ToDisplayString(HtmlFormatter.MimeType);

            formatted.Should()
                .Contain("'https://a.cdn.url/script'")
                .And
                .NotContain("'https://a.cdn.url/script.js'");
        }

        [Fact]
        public async Task can_specify_cacheBuster()
        {
            using var kernel = new CompositeKernel();

            var kernelExtension = new SandDanceKernelExtension();
            await kernelExtension.OnLoadAsync(kernel);

            var data = new[]
            {
                new {Type="orange", Price=1.2},
                new {Type="apple" , Price=1.3},
                new {Type="grape" , Price=1.4}
            };

            var explorer = data.ExploreWithSandDance();
            explorer.LibraryUri = new Uri("https://a.cdn.url/script.js");
            explorer.CacheBuster = "XYZ";
            var formatted = explorer.ToDisplayString(HtmlFormatter.MimeType);

            formatted.Should().Contain("'urlArgs': 'cacheBuster=XYZ'");
        }

        public void Dispose()
        {
            Formatter.ResetToDefault();
            SandDanceDataExplorer.ResetDefaultConfiguration();
        }
    }
}