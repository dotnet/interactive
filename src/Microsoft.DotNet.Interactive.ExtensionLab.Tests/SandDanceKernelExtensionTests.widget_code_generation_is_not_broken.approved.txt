<div style="background-color:white;">
<div id="00000000000000000000000000000000" style="height: 100ch ;margin: 2px;">
</div>
<script type="text/javascript">

renderSandDanceExplorer_00000000000000000000000000000000 = () => {

    (require.config({ 'paths': { 'context': '1.0.252001', 'sandDanceUri' : 'https://colombod.github.io/dotnet-interactive-cdn/extensionlab/1.0.252001/SandDance/sanddanceapi', 'urlArgs': 'cacheBuster=00000000000000000000000000000000' }}) || require)(['sandDanceUri'], (sandDance) => {

        sandDance.createSandDanceExplorer({
            data: {
  "profile": "tabular-data-resource",
  "schema": {
    "primaryKey": [],
    "fields": [
      {
        "name": "Type",
        "type": "string"
      },
      {
        "name": "Price",
        "type": "number"
      }
    ]
  },
  "data": [
    {
      "Type": "orange",
      "Price": 1.2
    },
    {
      "Type": "apple",
      "Price": 1.3
    },
    {
      "Type": "grape",
      "Price": 1.4
    }
  ]
},
            id: "00000000000000000000000000000000",
            container: document.getElementById("00000000000000000000000000000000")
        });
    },
    (error) => {
        console.log(error);
    });
};
// ensure `require` is available globally
if ((typeof(require) !==  typeof(Function)) || (typeof(require.config) !== typeof(Function))) {
    let require_script = document.createElement('script');
    require_script.setAttribute('src', 'https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js');
    require_script.setAttribute('type', 'text/javascript');
    
    
    require_script.onload = function() {
        renderSandDanceExplorer_00000000000000000000000000000000();
    };

    document.getElementsByTagName('head')[0].appendChild(require_script);
}
else {
    renderSandDanceExplorer_00000000000000000000000000000000();
}

 </script>
</div>
